### ISSUE TITLE
Assess use of profile-guided optimization

### ISSUE THREAD

**Issue by fako1024** [els0r/goProbe#138]
As of Go 1.21 profile-guided optimizations are supported (see https://tip.golang.org/doc/pgo for details), which means that the compiler is able to use production-level profiles to improve performance based on hot paths identified from them. This could be very interesting for both `goProbe` and `goQuery` (in particular the latter due to the simplicity to gather profiles on any host, not only a PROD host and because I gather the complex processing happening in a long query might actually benefit from PGO).

DoD

[ ] Gather some profiles from large queries locally
[ ] Re-compile using PGO
[ ] Check for performance improvements

**Comment by els0r** [comment]
Wow! I can help produce profiles of different flavors (many interfaces, much traffic, etc.).

**Comment by fako1024** [comment]
@els0r I've setup a first shot using the latest profile (for goprobe only) on the linked branch. You should be able to compile using `go build -pgo=auto` in the `cmd/goProbe` directory to create a binary with PGO enabled (Go 1.20 and up). If that does something measurable though I do not know...

**Comment by fako1024** [comment]
In order to get a feeling I've run `goProbe` twice in parallel from this branch for a couple of hours on my local device, once with `go build -pgo=auto` and once with `go build -pgo=off`. Start was exactly at the same time, same config (obviously). Interestingly, startup seemed to be a bit more expensive (for half an hour or so, non-PGO was leading the "race" upfront), but after a while PGO "overtook" and consistently stayed ahead:
```
└─ $ ▶ top -b -n1 | grep goProbe
 162370 root      20   0 1781716 257140 249216 S   0.0   1.6   0:03.67 goProbe
 162371 root      20   0 1855480 257052 248960 S   0.0   1.6   0:03.51 goProbe_pgo
```
The difference is about 4.4% cumulative, which is well within the margin of expected improvements using PGO (2-7%). Of course this is a tiny sample with only one small production profile as source for PGO, but since it's (almost) free: why not. goQuery might be even more interesting of course... :stuck_out_tongue:

**Comment by fako1024** [comment]
@els0r Is there any chance that we pull another set of fresh CPU profiles on the four hosts in https://github.com/els0r/goProbe/issues/74 ? Rationale(s):

- I only got profiles for two of them so far
- We have added some significant changes recently (on the slimcap side), only one of the two profiles has those
- If we are confident about the current version we could maybe run the profiles longer than the last time to increase the information available to the profiler
- We probably want to run further tests on these four hosts anyway as part of the aforementioned issue, so it's probably a nul-op

What do you think?

**Comment by fako1024** [comment]
@els0r FYI: I've re-added the goQuery steps to the ToDo list for the release (and opened https://github.com/els0r/goProbe/issues/170 for later automation) because I'd like to at least get a very basic `default.pgo` in once https://github.com/els0r/goProbe/issues/168 is merged (which I'll capture locally on my machine, good enough for a first step and basically zero-effort). That way we can get an idea if that's of any use at all and I can also update the build pipeline to use PGO.

**Comment by els0r** [comment]
Thanks! Makes sense. I'm a bit swamped with the different production test scenarios for goprobe. Would appreciate if you can run against your 10y+ DB and collect the first profile from there.

I'm happy to supply more profiles when we have the larger DBs in production as part of continuous profiling/optimization.

**Comment by fako1024** [comment]
> Would appreciate if you can run against your 10y+ DB and collect the first profile from there.

Absolutely, that was the plan. I think _something_ is better than nothing (at least that's what I hope), we can improve on this later...

**Comment by fako1024** [comment]
Ran another ~24h test on my machine, no big surprises here:
```
 123610 root       249600 S   0.0   1.6   0:26.04 goProbe_nopgo
 123611 root       249472 S   0.0   1.6   0:25.91 ./goProbe_pgo
```
Not really a big gain this time around TBH, but well - the profiles don't really match my system, and it's a really boring setup.

**Comment by fako1024** [comment]
Obtained a basic `default.pgo` from a couple of samples of my biggest DB using the following (crude) script:
```
#!/bin/bash

N=10

rm -rf /tmp/testprof

for i in `seq 1 $N`; do
 ./goQuery -d /tmp/db_out2 -f '28.6.01 00:00' sip -i eth0 --profiling.output-dir /tmp/testprof/${i}/1
 ./goQuery -d /tmp/db_out2 -f '28.6.01 00:00' dip -i eth0 --profiling.output-dir /tmp/testprof/${i}/2
 ./goQuery -d /tmp/db_out2 -f '28.6.01 00:00' dport -i eth0 --profiling.output-dir /tmp/testprof/${i}/3
 ./goQuery -d /tmp/db_out2 -f '28.6.01 00:00' proto -i eth0 --profiling.output-dir /tmp/testprof/${i}/4
 ./goQuery -d /tmp/db_out2 -f '28.6.01 00:00' sip,dip,dport,proto -i eth0 --profiling.output-dir /tmp/testprof/${i}/5
 ./goQuery -d /tmp/db_out2 -f '28.6.01 00:00' sip,dip -i eth0 --profiling.output-dir /tmp/testprof/${i}/6
 ./goQuery -d /tmp/db_out2 -f '28.6.01 00:00' dport,proto -i eth0 --profiling.output-dir /tmp/testprof/${i}/7
done

go tool pprof -proto /tmp/testprof/*/*/goquery_cpu_profile.pprof > /tmp/testprof/merged.pprof

mv /tmp/testprof/merged.pprof ./default.pgo
```

A test compilation shows no significant improvement. Maybe this will improve over time when using a) more representative samples and b) PGO gets better.
